gem/impero_utils.py

```diff
@@ -241,7 +241,7 @@ def recurse_block(expr, loop_indices):
             refcount.update(recurse(statement, loop_indices))
 
         # Visit :class:`collections.Counter` in deterministic order
-        for e in sorted(refcount.keys(), key=temporaries.index):
```

Comment:
`temporaries` is a `list`, `list.index` is linear search, which is called to get the key of every temporary.